stl/inc/xutility

```diff
@@ -4405,12 +4405,39 @@ _OutIt _Copy_unchecked(_InIt _First, _InIt _Last, _OutIt _Dest) {
 }
 #endif // _HAS_IF_CONSTEXPR
 
+// _Iterators_are_contiguous<_Iter1, _Iter2> reports whether both iterators are known to be contiguous.
+// (Without concepts, this detection is limited, which will limit when we can activate the memcmp optimization.)
+
+#ifdef __cpp_lib_concepts
+// When concepts are available, we can detect arbitrary contiguous iterators.
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = contiguous_iterator<_Iter1> //
+    && contiguous_iterator<_Iter2>;
+#else // ^^^ defined(__cpp_lib_concepts) ^^^ / vvv !defined(__cpp_lib_concepts) vvv
+// When concepts aren't available, we can detect pointers. (Iterators should be unwrapped before using this.)
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>>;
+#endif // ^^^ !defined(__cpp_lib_concepts) ^^^
+
 template <class _InIt, class _OutIt>
 _CONSTEXPR20 _OutIt copy(_InIt _First, _InIt _Last, _OutIt _Dest) { // copy [_First, _Last) to [_Dest, ...)
     _Adl_verify_range(_First, _Last);
     const auto _UFirst = _Get_unwrapped(_First);
     const auto _ULast  = _Get_unwrapped(_Last);
     const auto _UDest  = _Get_unwrapped_n(_Dest, _Idl_distance<_InIt>(_UFirst, _ULast));
+
+#if _ITERATOR_DEBUG_LEVEL == 2
+    if _CONSTEXPR_IF (_Iterators_are_contiguous<_Unwrapped_t<_InIt>, _Unwrapped_t<_OutIt>>) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            _STL_VERIFY(
+                _UDest + (_ULast - _UFirst) <= _UFirst || _ULast <= _UDest, "ranges should not overlap each other");

```

Comment:
The tests are failing with:
```
D:\build\x86\out\inc\xutility(4435): error C2676: binary '+': 'const _OutIt' does not define this operator or a conversion to a type acceptable to the predefined operator
        with
        [
            _OutIt=std::ostreambuf_iterator<wchar_t,std::char_traits<wchar_t>>
        ]
```
This is because when `if constexpr` isn't available, the `if _CONSTEXPR_IF` above is just a plain `if`, and this code is compiled for non-contiguous iterators.

We'll be able to use `if constexpr` unconditionally after resolving #639 but that hasn't happened yet.

As a workaround, it would be reasonable to guard these debug checks with `#if _HAS_IF_CONSTEXPR` and say `if constexpr` within. That will be active for all of our supported front-ends except CUDA (and the tests that intentionally disable this). Such a guard would be reasonable because these are only debug checks and aren't needed for conformance, so we don't need to write a tag dispatch fallback.

Reply:
Thank you. Your hint Saved me a lot of time

Reply:
Thank you. Your hint saved me a lot of time.